pkg/ccl/changefeedccl/parquet_sink_cloudstorage.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2019 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package changefeedccl
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdcevent"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent"
+	pqexporter "github.com/cockroachdb/cockroach/pkg/sql/importer"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
+	goparquet "github.com/fraugster/parquet-go"
+	"github.com/fraugster/parquet-go/parquetschema"
+)
+
+// This variable is used to make sure that we add primary keys of the table to
+// the metadata of the parquet file only under testing (The primary keys are
+// used to convert the parquet data into JSON)
+var includeParquetTestMetadata = false
+
+// We need a separate sink for parquet format because the parquet encoder has to
+// write metadata to the parquet file (buffer) after each flush. This means that the
+// parquet encoder should have access to the buffer object inside
+// cloudStorageSinkFile file. This means that the parquet writer has to be
+// embedded in the cloudStorageSinkFile file. If we wanted to maintain the
+// existing separation between encoder and the sync, then we would need to
+// figure out a way to get the embedded parquet writer in the
+// cloudStorageSinkFile and pass it to the encode function in the encoder.
+// Instead of this it logically made sense to have a single sink for parquet
+// format which did the job of both encoding and emitting to the cloud storage.
+// This sink currently embeds the cloudStorageSinkFile and it has a function
+// EncodeAndEmitRow which links the buffer in the cloudStorageSinkFile and the
+// parquetWriter every time we need to create a file for each unique combination
+// of topic and schema version (We need to have a unique parquetWriter for each
+// unique combination of topic and schema version because each parquet file can
+// have a single schema written inside it and each parquetWriter can only be
+// associated with a single schema.)
+type parquetCloudStorageSink struct {
+	wrapped *cloudStorageSink
+}
+
+type parquetWriterWrapper struct {
```

Comment:
this struct only writes one parquet file, correct? It seems more descriptive to call it a `parquetFileWriter`, i think?